cpp/include/raft/comms/std_comms.hpp

```diff
@@ -0,0 +1,406 @@
+      printf("NCCL call='%s' failed. Reason:%s\n", #call, \
+             ncclGetErrorString(status));                 \
+    }                                                     \
+  } while (0)
+
+namespace raft {
+namespace comms {
+
+size_t getDatatypeSize(const datatype_t datatype) {
+  switch (datatype) {
+    case CHAR:
+      return sizeof(char);
+    case UINT8:
+      return sizeof(uint8_t);
+    case INT:
+      return sizeof(int);
+    case UINT:
+      return sizeof(unsigned int);
+    case INT64:
+      return sizeof(int64_t);
+    case UINT64:
+      return sizeof(uint64_t);
+    case FLOAT:
+      return sizeof(float);
+    case DOUBLE:
+      return sizeof(double);
+  }
+}
+
+ncclDataType_t getNCCLDatatype(const datatype_t datatype) {
+  switch (datatype) {
+    case CHAR:
+      return ncclChar;
+    case UINT8:
+      return ncclUint8;
+    case INT:
+      return ncclInt;
+    case UINT:
+      return ncclUint32;
+    case INT64:
+      return ncclInt64;
+    case UINT64:
+      return ncclUint64;
+    case FLOAT:
+      return ncclFloat;
+    case DOUBLE:
+      return ncclDouble;
+  }
+}
+
+ncclRedOp_t getNCCLOp(const op_t op) {
+  switch (op) {
+    case SUM:
+      return ncclSum;
+    case PROD:
+      return ncclProd;
+    case MIN:
+      return ncclMin;
+    case MAX:
+      return ncclMax;
+  }
+}
+
+class std_comms : public comms_iface {
+ public:
+  std_comms() = delete;
+
+  /**
+   * @brief Constructor for collective + point-to-point operation.
+   * @param comm initialized nccl comm
+   * @param ucp_worker initialized ucp_worker instance
+   * @param eps shared pointer to array of ucp endpoints
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t comm, ucp_worker_h ucp_worker,
+            std::shared_ptr<ucp_ep_h *> eps, int size, int rank)
+    : _nccl_comm(comm),
+      _ucp_worker(ucp_worker),
+      _ucp_eps(eps),
+      _size(size),
+      _rank(rank),
+      _next_request_id(0) {
+    initialize();
+    p2p_enabled = true;
+  };
+
+  /**
+   * @brief constructor for collective-only operation
+   * @param comm initilized nccl communicator
+   * @param size size of the cluster
+   * @param rank rank of the current worker
+   */
+  std_comms(ncclComm_t comm, int size, int rank)
+    : _nccl_comm(comm), _size(size), _rank(rank) {
+    initialize();
+  };
+
+  virtual ~std_comms() {
+    CUDA_CHECK_NO_THROW(cudaStreamDestroy(_stream));
+
+    CUDA_CHECK_NO_THROW(cudaFree(_sendbuff));
+    CUDA_CHECK_NO_THROW(cudaFree(_recvbuff));
+  }
+
+  void initialize() {
+    CUDA_CHECK(cudaStreamCreate(&_stream));
+
+    CUDA_CHECK(cudaMalloc(&_sendbuff, sizeof(int)));
+    CUDA_CHECK(cudaMalloc(&_recvbuff, sizeof(int)));
+  }
+
+  int getSize() const { return _size; }
+
+  int getRank() const { return _rank; }
+
+  std::unique_ptr<comms_iface> commSplit(int color, int key) const {
+    // Not supported by NCCL
```

Comment:
Interesting. Sylvain always told me to avoid using more than one communicator in the same process (Eg same gpu) because it can lead to deadlocks when the streams are not able to schedule warps concurrently. Perhaps this changed in more recent versions? We might want to investigate this a little further to be sure.

Reply:
It may be problematic if we concurrently run multiple communicators (ideally this should work as well and this works pretty well for MPI but this may not be a case for NCCL yet), but if we communicate using different communicators in different phases, I haven't encountered any problem.

Reply:
Since you brought it up, I also want to point out that my other concern over this feature was related to high initialization latencies I was experiencing when creating new NCCL communicators. Contrary to the expected behavior, these latencies were not being reduced at all in subsequent initializations.

@divyegala noticed in some of his recent benchmarking that this appears to have been fixed.

Reply:
Yeah... I've seen that issue as well. But IIRC, this overhead gets worse with process count, so creating sub-communicators is  generally cheaper than creating a full communicator.

Nonetheless no need to pay unnecessary overhead if sub-communicators are unnecessary, and we may better not initialize sub-communicators by default but better provide an option.